src/ripple/overlay/impl/MessageScheduler.cpp

```diff
@@ -0,0 +1,413 @@
+    {
+        push_value top(caller, "schedule");
+        _offer(channels_, senders);
+    }
+    std::move(senders.begin(), senders.end(), std::back_inserter(senders_));
+    return true;
+}
+
+RequestId
+MessageScheduler::send(
+    std::shared_ptr<Peer> peer,
+    protocol::TMGetLedger& message,
+    Receiver* receiver,
+    NetClock::duration timeout)
+{
+    auto requestId = ++nextId_;
+    message.set_requestcookie(requestId);
+    _send(peer, requestId, message, receiver, timeout);
+    return requestId;
+}
+
+RequestId
+MessageScheduler::send(
+    std::shared_ptr<Peer> peer,
+    protocol::TMGetObjectByHash& message,
+    Receiver* receiver,
+    NetClock::duration timeout)
+{
+    auto requestId = ++nextId_;
+    message.set_seq(requestId);
+    JLOG(journal_.trace()) << "send,type=get_objects,count="
+                           << message.objects_size();
+    _send(peer, requestId, message, receiver, timeout);
+    return requestId;
+}
+
+void
+MessageScheduler::_offer(
+    std::vector<Channel>& channels,
+    std::vector<Sender*>& senders)
+{
+    JLOG(journal_.trace()) << "offer,during=" << caller
+                           << ",channels=" << channels.size()
+                           << ",senders=" << senders.size();
+    assert(!channels.empty());
+    assert(!senders.empty());
+    push_value top1(tsenders, &senders);
+    // We must iterate `senders` using an index
+    // because the offer may grow the vector,
+    // invalidating any iterators or references.
+    for (auto i = 0; i < senders.size(); ++i)
+    {
+        // If this is the last sender, offer it the full set of channels.
+        // If there are more senders waiting, offer one at a time, in turn.
+        auto size = (i + 1 == senders.size()) ? channels.size() : 1;
+        {
+            Offer offer{*this, channels, size};
+            try
+            {
+                senders[i]->onOffer(offer);
+            }
+            catch (std::exception const&)
+            {
+            }
+            if (offer.closed())
+            {
+                senders[i] = nullptr;
+            }
+            // Closed channels are removed here,
+            // in the destructor of the offer.
+        }
+        if (channels.empty())
+        {
+            break;
+        }
+    }
+    auto end = std::remove(senders.begin(), senders.end(), nullptr);
+    senders.erase(end, senders.end());
+}
+
+void
+MessageScheduler::_send(
+    std::shared_ptr<Peer> peer,
+    RequestId requestId,
+    ::google::protobuf::Message const& message,
+    protocol::MessageType type,
+    Receiver* receiver,
+    NetClock::duration timeout)
+{
+    // TODO: Add a factory method to Message for this pattern.
+    // It will let us remove the template function `_send`.
+    auto packet = std::make_shared<Message>(message, type);
+    // C++20: Switch to make_shared<Request>(...).
+    // https://stackoverflow.com/a/55144743/618906
+    auto request = std::unique_ptr<Request>(new Request{
```

Comment:
You can also use `make_unique` here.

Reply:
@drlongle I don't think you can. The `Request` constructor isn't accessible from `make_unique`.